xpart/particles/particles.py

```diff
@@ -157,109 +159,346 @@ class Particles(xo.HybridClass):
+            kwargs.pop(field, None)
 
-        if 'start_tracking_at_element' not in kwargs.keys():
-            self.start_tracking_at_element = -1
+        for xotype, field in per_part_input_vars:
+            if field not in kwargs.keys():
+                continue
 
-        # Initialize coordinates
-        with self._bypass_linked_vars():
-            if pyparticles is not None:
-                context = self._buffer.context
-                for tt, kk in list(scalar_vars):
-                    setattr(self, kk, part_dict[kk])
-                for tt, kk in list(per_particle_vars):
-                    if kk.startswith('_rng'):
-                        getattr(self, kk)[:] = 0
-                        continue
-                    vv = getattr(self, kk)
-                    vals =  context.nparray_to_context_array(part_dict[kk])
-                    ll = len(vals)
-                    vv[:ll] = vals
-                    vv[ll:] = LAST_INVALID_STATE
+            if np.isscalar(kwargs[field]) or len(kwargs[field]) == 1:
+                value = np.array(kwargs[field]).item()
+                kwargs[field] = np.full(num_particles, value)
             else:
-                for tt, kk in list(scalar_vars):
-                    setattr(self, kk, 0.)
-
-                for tt, kk in list(per_particle_vars):
-                    if kk == 'chi' or kk == 'charge_ratio' or kk == 'state':
-                        value = 1.
-                    elif kk == 'particle_id':
-                        value = np.arange(0, self._capacity, dtype=np.int64)
-                    else:
-                        value = 0.
-                    getattr(self, kk)[:] = value
-
-            self._num_active_particles = -1  # To be filled in only on CPU
-            self._num_lost_particles = -1  # To be filled in only on CPU
-
-            # Force values provided by user if compatible
-            for nn in part_energy_varnames() + ['p0c', 'beta0', 'gamma0']:
-                vvv = self._buffer.context.nparray_from_context_array(getattr(self, nn))
-                if nn in input_kwargs.keys():
-                    if hasattr(input_kwargs[nn], '__len__'):
-                        ll = len(input_kwargs[nn]) # in case there is unallocated space
-                    else:
-                        ll = len(vvv)
-
-                    if np.isscalar(input_kwargs[nn]):
-                        getattr(self, nn)[:] = input_kwargs[nn]
-                    else:
-                        getattr(self, nn)[:ll] = (
-                                context.nparray_to_context_array(
-                                    np.array(input_kwargs[nn])))
+                kwargs[field] = np.array(kwargs[field])
+
+            # coerce the right type so that we can allocate the right array
+            # in the target context. PyOpenCL gets fussy if types don't match
+            # in calculations.
+            kwargs[field] = kwargs[field].astype(xotype._dtype)
```

Comment:
Let's only do it when the type doesn't match (to explicitly avoid copies the possible)